File: Gestion_stock_kouttab_react/frontend/src/lib/naming.ts
/**
 * Nomenclature des fichiers envoyés au service comptable.
 *
 * Format : `{Pôle}_{Événement}_{AAAA-MM-JJ}.pdf`
 *
 * ⚠️ Ce module duplique volontairement `backend/app/services/naming.py`.
 * Le backend reste seul juge du nom réellement produit ; cette copie sert
 * uniquement à montrer au déposant, avant validation, le nom que portera sa
 * pièce. Les deux implémentations sont couvertes par la **même table de cas**
 * (`naming.test.ts` / `test_naming.py`) : toute divergence casse un test.
 */

export const MAX_COMPONENT_LEN = 40;
export const MAX_STEM_LEN = 120;
export const MISSING = 'NC';

/** Caractères que la normalisation NFKD ne décompose pas. */
const EXPLICIT_REPLACEMENTS: Array<[string, string]> = [
  ['œ', 'oe'],
  ['Œ', 'OE'],
  ['æ', 'ae'],
  ['Æ', 'AE'],
  ['ß', 'ss'],
  ['€', 'EUR'],
  ['$', 'USD'],
  ['£', 'GBP'],
  ['&', 'et'],
  ['@', 'at'],
  ['°', 'deg'],
  ['’', ''],
  ["'", ''],
  ['`', ''],
  ['"', ''],
  ['«', ''],
  ['»', ''],
];

/** Un fichier nommé `CON.pdf` est indécompressable sous Windows. */
const WINDOWS_RESERVED = new Set([
  'CON',
  'PRN',
  'AUX',
  'NUL',
  ...Array.from({ length: 9 }, (_, i) => `COM${i + 1}`),
  ...Array.from({ length: 9 }, (_, i) => `LPT${i + 1}`),
]);

export function slugifyComponent(
  value: string | null | undefined,
  maxLen = MAX_COMPONENT_LEN,
): string {
  if (value === null || value === undefined) return MISSING;

  let text = String(value);
  for (const [from, to] of EXPLICIT_REPLACEMENTS) {
    text = text.split(from).join(to);
  }

  // Décomposition puis suppression des diacritiques : é -> e, ç -> c.
  text = text.normalize('NFKD').replace(/[̀-ͯ]/g, '');
  // Filet de sécurité : alphabets non translittérables (cyrillique, CJK…).
  // Équivalent de `encode('ascii', 'ignore')` côté Python. Écrit sans regex :
  // une classe de caractères couvrant \x00 déclenche la règle no-control-regex.
  text = Array.from(text)
    .filter((ch) => ch.charCodeAt(0) < 128)
    .join('');

  // Neutralise d'un coup les interdits Windows, les caractères de contrôle,
  // les espaces et les points.
  text = text.replace(/[^A-Za-z0-9]+/g, '-');
  text = text.replace(/-{2,}/g, '-').replace(/^-+|-+$/g, '');

  if (!text) return MISSING;

  if (text.length > maxLen) {
    text = text.slice(0, maxLen);
    const lastDash = text.slice(1).lastIndexOf('-');
    if (lastDash !== -1) text = text.slice(0, lastDash + 1);
    text = text.replace(/^-+|-+$/g, '');
  }

  if (!text) return MISSING;
  if (WINDOWS_RESERVED.has(text.toUpperCase())) return `_${text}`;
  return text;
}

/** Date au format AAAA-MM-JJ. Accepte une chaîne de `<input type="date">`. */
export function formatDateComponent(value: string | Date | null | undefined): string {
  if (!value) return MISSING;
  if (typeof value === 'string') {
    const match = value.match(/^(\d{4})-(\d{2})-(\d{2})/);
    return match ? match[0] : MISSING;
  }
  if (Number.isNaN(value.getTime())) return MISSING;
  const year = value.getFullYear();
  const month = `${value.getMonth() + 1}`.padStart(2, '0');
  const day = `${value.getDate()}`.padStart(2, '0');
  return `${year}-${month}-${day}`;
}

export function buildAttachmentStem(
  components: Array<string | null | undefined>,
  dateValue: string | Date | null | undefined,
): string {
  const parts = components.map((c) => slugifyComponent(c));
  parts.push(formatDateComponent(dateValue));
  let stem = parts.join('_');
  if (stem.length > MAX_STEM_LEN) {
    stem = stem.slice(0, MAX_STEM_LEN).replace(/^[-_]+|[-_]+$/g, '');
  }
  return stem;
}

export function buildAttachmentFilename(
  components: Array<string | null | undefined>,
  dateValue: string | Date | null | undefined,
  extension = 'pdf',
): string {
  const ext = extension.replace(/^\./, '').toLowerCase() || 'pdf';
  return `${buildAttachmentStem(components, dateValue)}.${ext}`;
}

/**
 * Rend uniques des noms identiques en suffixant `-2`, `-3`…
 *
 * Les justificatifs d'un même dépôt partagent pôle, événement et date : sans
 * cette passe, les pièces jointes porteraient toutes le même nom.
 */
export function deduplicateFilenames(names: string[]): string[] {
  const seen = new Map<string, number>();
  return names.map((name) => {
    const dot = name.lastIndexOf('.');
    let stem = dot > 0 ? name.slice(0, dot) : name;
    const ext = dot > 0 ? name.slice(dot + 1) : '';

    const key = name.toLowerCase();
    const count = (seen.get(key) ?? 0) + 1;
    seen.set(key, count);
    if (count === 1) return name;

    const suffix = `-${count}`;
    // Laisse la place au suffixe : sinon la troncature le supprimerait et la
    // collision réapparaîtrait silencieusement.
    if (stem.length + suffix.length > MAX_STEM_LEN) {
      stem = stem.slice(0, MAX_STEM_LEN - suffix.length).replace(/^[-_]+|[-_]+$/g, '');
    }
    return ext ? `${stem}${suffix}.${ext}` : `${stem}${suffix}`;
  });
}

File: Gestion_stock_kouttab_react/backend/app/services/naming.py
"""Nomenclature des fichiers envoyes au service comptable.

Format retenu avec le client : ``{Pole}_{Evenement}_{AAAA-MM-JJ}.pdf``

Deux conventions de separateur, volontairement distinctes :
- ``_`` separe les composants,
- ``-`` remplace tout caractere invalide *a l'interieur* d'un composant.

Le comptable peut ainsi re-decouper un nom par ``split("_")`` sans ambiguite,
ce qui serait impossible si les deux roles se confondaient.

La sortie est garantie **ASCII pur** : aucun encodage RFC 2231 n'est necessaire
dans l'en-tete ``Content-Disposition``, donc aucun client de messagerie ne
deforme le nom de la piece jointe.

Module volontairement pur : aucune I/O, aucun acces base, aucune dependance a
FastAPI. Il est donc testable exhaustivement (cf. ``tests/unit/test_naming.py``)
et duplique a l'identique cote frontend (``src/lib/naming.ts``) pour afficher au
deposant le nom exact qui sera envoye.
"""

from __future__ import annotations

import re
import unicodedata
from collections.abc import Sequence
from datetime import date


MAX_COMPONENT_LEN = 40
MAX_STEM_LEN = 120
MISSING = "NC"
DATE_FORMAT = "%Y-%m-%d"

# Caracteres que la decomposition NFKD ne separe pas : ils disparaitraient
# purement et simplement a l'etape ASCII sans ce passage prealable.
_EXPLICIT_REPLACEMENTS = {
    "œ": "oe",
    "Œ": "OE",
    "æ": "ae",
    "Æ": "AE",
    "ß": "ss",
    "€": "EUR",
    "$": "USD",
    "£": "GBP",
    "&": "et",
    "@": "at",
    "°": "deg",
    "'": "",
    "'": "",
    "`": "",
    '"': "",
    "«": "",
    "»": "",
}

# Noms reserves par Windows : un fichier « CON.pdf » est indecompressable et
# provoque des erreurs opaques cote destinataire.
_WINDOWS_RESERVED = {
    "CON", "PRN", "AUX", "NUL",
    *(f"COM{i}" for i in range(1, 10)),
    *(f"LPT{i}" for i in range(1, 10)),
}

_NON_ALNUM = re.compile(r"[^A-Za-z0-9]+")
_MULTI_DASH = re.compile(r"-{2,}")


def slugify_component(value: str | None, *, max_len: int = MAX_COMPONENT_LEN) -> str:
    """Reduit un libelle libre a un composant de nom de fichier sur.

    Retourne ``MISSING`` si rien d'exploitable ne subsiste — jamais une chaine
    vide, qui produirait un ``__`` cassant le decoupage du nom.
    """
    if value is None:
        return MISSING

    text = str(value)
    for source, target in _EXPLICIT_REPLACEMENTS.items():
        text = text.replace(source, target)

    # Decomposition puis suppression des diacritiques : e -> e, c -> c.
    text = unicodedata.normalize("NFKD", text)
    text = "".join(ch for ch in text if not unicodedata.combining(ch))
    # Filet de securite pour les alphabets non translitterables (cyrillique,
    # arabe, CJK) : mieux vaut les perdre que produire des octets invalides.
    text = text.encode("ascii", "ignore").decode("ascii")

    # Neutralise d'un coup les interdits Windows (< > : " / \ | ? *), les
    # caracteres de controle, les espaces et les points.
    text = _NON_ALNUM.sub("-", text)
    text = _MULTI_DASH.sub("-", text).strip("-")

    if not text:
        return MISSING

    if len(text) > max_len:
        text = text[:max_len]
        # Coupe sur une frontiere de mot plutot qu'au milieu, si possible.
        if "-" in text[1:]:
            text = text[: text.rindex("-")]
        text = text.strip("-")

    if not text:
        return MISSING
    if text.upper() in _WINDOWS_RESERVED:
        text = f"_{text}"
    return text


def format_date_component(value: date | None) -> str:
    return value.strftime(DATE_FORMAT) if value is not None else MISSING


def build_attachment_stem(
    components: Sequence[str | None],
    date_value: date | None,
) -> str:
    """Assemble le nom (sans extension) a partir des composants et de la date."""
    parts = [slugify_component(c) for c in components]
    parts.append(format_date_component(date_value))
    stem = "_".join(parts)
    if len(stem) > MAX_STEM_LEN:
        stem = stem[:MAX_STEM_LEN].strip("-_")
    return stem


def build_attachment_filename(
    components: Sequence[str | None],
    date_value: date | None,
    *,
    extension: str = "pdf",
) -> str:
    ext = extension.lstrip(".").lower() or "pdf"
    return f"{build_attachment_stem(components, date_value)}.{ext}"


def deduplicate_filenames(names: Sequence[str]) -> list[str]:
    """Rend uniques des noms identiques en suffixant ``-2``, ``-3``...

    Les justificatifs d'un meme depot partagent pole, evenement et date : sans
    cette passe, les N pieces jointes porteraient le meme nom et les clients de
    messagerie en ecraseraient certaines.

    Pas de suffixe entre parentheses facon « (1) » : les espaces et parentheses
    survivent mal aux anciennes passerelles SMTP.
    """
    seen: dict[str, int] = {}
    result: list[str] = []
    for name in names:
        stem, _, ext = name.rpartition(".")
        if not stem:  # nom sans extension
            stem, ext = name, ""
        key = name.lower()
        count = seen.get(key, 0) + 1
        seen[key] = count
        if count == 1:
            result.append(name)
            continue
        suffix = f"-{count}"
        # Laisse la place au suffixe, sinon la troncature le ferait sauter et la
        # collision reapparaitrait silencieusement.
        if len(stem) + len(suffix) > MAX_STEM_LEN:
            stem = stem[: MAX_STEM_LEN - len(suffix)].strip("-_")
        result.append(f"{stem}{suffix}.{ext}" if ext else f"{stem}{suffix}")
    return result
